Clarify async scenarios and deadlocks

adegeo opened the pull request:
## Summary

A long time coming. Updates a few places with information about deadlocks and demonstrates code. Also adds reference to the new InvokeAsync methods in .NET 9.

Fixes #1406


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md](https://github.com/dotnet/docs-desktop/blob/88e17acc86144f61d2b840a5414fcabb07918a86/dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md) | [dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls](https://review.learn.microsoft.com/en-us/dotnet/desktop/winforms/controls/how-to-make-thread-safe-calls?branch=pr-en-us-2112) |
| [dotnet-desktop-guide/winforms/forms/events.md](https://github.com/dotnet/docs-desktop/blob/88e17acc86144f61d2b840a5414fcabb07918a86/dotnet-desktop-guide/winforms/forms/events.md) | [dotnet-desktop-guide/winforms/forms/events](https://review.learn.microsoft.com/en-us/dotnet/desktop/winforms/forms/events?branch=pr-en-us-2112) |


<!-- PREVIEW-TABLE-END -->

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md:
```diff
@@ -61,3 +71,62 @@ The example counts from 0 to 10 in the `DoWork` event, pausing for one second be
+
+- Synchronously sends the delegate to the UI thread's message queue.
+- The calling thread waits until the UI thread processes the delegate.
+- Can lead to UI freezes if overused during long-running operations.
```
```suggestion
- Can lead to UI freezes, when the delegate, which was marshalled to the message queue, is itself waiting for a certain message to arrive (dead lock).
```

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md:
```diff
@@ -61,3 +71,62 @@ The example counts from 0 to 10 in the `DoWork` event, pausing for one second be
+- Synchronously sends the delegate to the UI thread's message queue.
+- The calling thread waits until the UI thread processes the delegate.
+- Can lead to UI freezes if overused during long-running operations.
+- Useful when immediate results are needed from the UI thread.
```
```suggestion
- Useful when already provided results are just needed to be picked-up from the UI thread (e.g. reading a control's property), or when a control does quickly need to be updated (updating just the Text properties of a handful of controls).
```

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md:
```diff
@@ -61,3 +71,62 @@ The example counts from 0 to 10 in the `DoWork` event, pausing for one second be
+**Control.InvokeAsync (Posting - Non-blocking):**
+
+- Asynchronously posts the delegate to the UI thread's message queue.
+- The calling thread doesn't wait and continues its work immediately.
```
```suggestion
- Rather than waiting for the Invoke to return in a blocking way, the calling thread schedules just a call-back, but returns immediately. Although the code flow seems to be waiting at the `await` - the calling thread remains responsive at all times, and the code flow continues automatically through that internally scheduled callback mechanism.
```

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md:
```diff
@@ -61,3 +71,62 @@ The example counts from 0 to 10 in the `DoWork` event, pausing for one second be
+- The calling thread doesn't wait and continues its work immediately.
+- Returns a `Task` that can be awaited for completion.
+- Ideal for async scenarios and prevents UI thread bottlenecks.
+
```
- Can also be used _from_ the UI-Thread, when you want a Message to be completely finished, before you want to run a specific method.

adegeo replied:
This is kind of implied with the first two points.

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md:
```diff
@@ -61,3 +71,62 @@ The example counts from 0 to 10 in the `DoWork` event, pausing for one second be
+| [`InvokeAsync(Func<CancellationToken, ValueTask>)`][invoke_func_value]               | Async operation, no return value.*   | Long-running UI updates.         |
+| [`InvokeAsync<T>(Func<CancellationToken, ValueTask<T>>)`][invoke1_func_value_return] | Async operation, with return value.* | Async data fetching with result. |
+
+*Visual Basic doesn't support using awaiting a <xref:System.Threading.Tasks.ValueTask>.
```
```suggestion
*Note that Visual Basic doesn't support the awaiting of asynchronous methods in methods which return <xref:System.Threading.Tasks.ValueTask>. In this case, please let your asynchronous methods return a Task and later wrap that Task in a ValueTask.
```

adegeo replied:
@KlausLoeffelmann  Can you be clearer about this? How does someone do that? It just seems to me that it's impossible to use the two "async operation" from VB because it's expecting a `ValueTask`, right? I must be missing something in connecting the dots 😁

KlausLoeffelmann replied:
So, when we say, VB cannot use `ValueTask`, that is not correct.
Visual Basic is limited in that the compiler cannot create the necessary state machines needed for compiling `await`-s _in the context of_ `ValueTask`.

But it does not mean, you cannot _use_ `ValueTask` in VB.

The following code is a series of extension methods, which converts a Task (with our required signatures for `InvokeAsync`, except they are not returning a `_Value_Task`) into the equivalent ValueTask:

```VisualBasic
Imports System.Runtime.CompilerServices
Imports System.Threading

Module ValueTaskExtensions

    ''' <summary>
    '''  Converts a Func(Of Task) into a Func(Of ValueTask).
    ''' </summary>
    <Extension>
    Public Function AsValueTask(work As Func(Of Task)) As Func(Of ValueTask)
        Return Function()
                   Return New ValueTask(work())
               End Function
    End Function

    ''' <summary>
    '''  Converts a Func(Of Task(Of T)) into a Func(Of ValueTask(Of T)).
    ''' </summary>
    <Extension>
    Public Function AsValueTask(Of T)(work As Func(Of Task(Of T))) As Func(Of ValueTask(Of T))
        Return Function()
                   Return New ValueTask(Of T)(work())
               End Function
    End Function

    ''' <summary>
    '''  Converts a Func(Of CancellationToken, Task) into a Func(Of CancellationToken, ValueTask).
    ''' </summary>
    <Extension>
    Public Function AsValueTask(work As Func(Of CancellationToken, Task)) As Func(Of CancellationToken, ValueTask)
        Return Function(ct As CancellationToken)
                   Return New ValueTask(work(ct))
               End Function
    End Function

    ''' <summary>
    '''  Converts a Func(Of CancellationToken, Task(Of T)) into a Func(Of CancellationToken, ValueTask(Of T)).
    ''' </summary>
    <Extension>
    Public Function AsValueTask(Of T)(work As Func(Of CancellationToken, Task(Of T))) As Func(Of CancellationToken, ValueTask(Of T))
        Return Function(ct As CancellationToken)
                   Return New ValueTask(Of T)(work(ct))
               End Function
    End Function
End Module
```

Now, if you have that, you can craft your async method to use in InvokeAsync like this:

```VisualBasic
    Private Async Function UpdateUi(cancellation As CancellationToken) As Task(Of String)

        Dim pi = Await PiGenerator.CalculateDigitsOfPiAsync(
            existingResult:=String.Empty,
            digits:=20,
            cancellationToken:=cancellation)

        Label1.Text = pi
        Return pi

    End Function
```

And now you can use the extension methods like this, to convert your Task<T> into ValueTask<T>:

```VisualBasic
    Private Async Sub BtnCalcPi_Click(sender As Object, e As EventArgs) Handles _btnCalcPi.Click
        Dim panel = New AutoColumnTableLayoutPanel
        Me.Controls.Add(panel)

        Dim func = DirectCast(
            AddressOf UpdateUi,
            Func(Of CancellationToken, Task(Of String))).AsValueTask()

        Await InvokeAsync(func, cancellation)

    End Sub
```

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/how-to-make-thread-safe-calls.md:
```diff
@@ -61,3 +71,62 @@ The example counts from 0 to 10 in the `DoWork` event, pausing for one second be
+
+### Advantages of InvokeAsync
+
+`Control.InvokeAsync` has several advantages over the older `Control.Invoke` method. It returns a `Task` that you can await, making it work well with async and await code. It also prevents common deadlock problems that can happen when mixing async code with synchronous invoke calls. Unlike `Control.Invoke`, the `InvokeAsync` method doesn't block your background thread, so your app stays more resp ...
```
```suggestion
`Control.InvokeAsync` has several advantages over the older `Control.Invoke` method. It returns a `Task` that you can await, making it work well with async and await code. It also prevents common deadlock problems that can happen when mixing async code with synchronous invoke calls. Unlike `Control.Invoke`, the `InvokeAsync` method doesn't block the calling thread, which keeps your apps responsive and avoids hangs.
```

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/cs/FormBackgroundWorker.cs:
```diff
@@ -8,37 +8,36 @@
+    private void button1_Click(object sender, EventArgs e)
+    {
+        if (!backgroundWorker1.IsBusy)
+            backgroundWorker1.RunWorkerAsync();
```
Should we add a comment that although this methods ends in `Async`, it does _not_ return a Task, thus cannot be awaited?

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/cs/FormThread.cs:
```diff
@@ -8,43 +8,42 @@
-namespace project
+namespace project;
+
+public partial class FormThread : Form
```
Shouldn't we ditch the Thread examples altogether?

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/cs/InvokeAsyncExamples.cs:
```diff
@@ -0,0 +1,108 @@
+
+namespace project;
+
+public partial class InvokeAsyncForm : Form
```
```suggestion
public partial class InvokeAsyncSampleForm : Form
```

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/controls/snippets/how-to-make-thread-safe-calls/cs/InvokeAsyncExamples.cs:
```diff
@@ -0,0 +1,108 @@
+                for (int i = 0; i <= 100; i += 10)
+                {
+                    // Simulate work
+                    await Task.Delay(100);
```
Maybe we're using something like Thread.Spin to actually simulate work?
Task.Delay doesn't do any "work".

KlausLoeffelmann commented on dotnet-desktop-guide/winforms/forms/events.md:
```diff
@@ -66,6 +67,82 @@ Typically each event produces an event handler with a different event-object typ
+:::code language="vb" source="snippets/events/vb/AsyncEventHandlers.vb" id="snippet_BasicAsyncEventHandler":::
+
+> [!IMPORTANT]
+> While `async void` is generally discouraged, it's necessary for event handlers since they cannot return `Task`. Always wrap awaited operations in `try-catch` blocks to handle exceptions properly, as shown in the example above.
```
```suggestion
> While `async void` is generally discouraged, it's the accepted pattern for event handler and event handler-like methods (when overriding one of the _event-raising_ methods, like `OnClick`). Make sure, though, to always wrap awaited operations in `try-catch` blocks to handle exceptions properly, as shown in the example above.
```

KlausLoeffelmann reviewed (changes requested):
Let's talk about a few aspects here.
We should also address `.ConfigureAwait(true/false)`, and point out, how it can help in simplifying certain type of marshalling work, but can also impose deadlocks when applied poorly.

For the `ValueTask` issue in VB, I think I have a workaround, which we should provide in the docs.

adegeo commented:
@KlausLoeffelmann I addressed your feedback.

- I also renamed snippets and code files to be clearer.
- I removed code that was unused.
- I also rewrote some of the snippets to better demonstrate the concepts being talked about here.
- One other thing I did was reorder the bullet points to emphasize InvokeAsync, then Invoke, then BackgroundWorker.

BillWagner reviewed (approved):
This looks really good @adegeo You can :shipit: when ready.
